README.md
### Tags and respective `Dockerfile` links

- [`2.5.0`](https://github.com/jdeathe/centos-ssh-varnish/releases/tag/2.5.0), `centos-7` [(centos-7/Dockerfile)](https://github.com/jdeathe/centos-ssh-varnish/blob/centos-7/Dockerfile)
- [`1.7.1`](https://github.com/jdeathe/centos-ssh-varnish/releases/tag/1.7.1), `centos-6` [(centos-6/Dockerfile)](https://github.com/jdeathe/centos-ssh-varnish/blob/centos-6/Dockerfile)

## Overview

This build uses the base image [jdeathe/centos-ssh](https://github.com/jdeathe/centos-ssh) so inherits it's features but with `sshd` disabled by default. [Supervisor](http://supervisord.org/) is used to start the varnishd (and optionally the varnishncsa) daemon when a docker container based on this image is run.

### Image variants

- [Varnish Cache 6.3 - CentOS-7](https://github.com/jdeathe/centos-ssh-varnish/blob/centos-7)
- [Varnish Cache 4.1 - CentOS-6](https://github.com/jdeathe/centos-ssh-varnish/blob/centos-6)

## Quick start

> For production use, it is recommended to select a specific release tag as shown in the examples.

Run up a container named `varnish.1` from the docker image `jdeathe/centos-ssh-varnish` on port 80 of your docker host. 1 backend host is defined mapping the host `httpd_1` to the IP address `172.17.8.101`; this is required to identify the backend host that's defined in the default Varnish VCL file. 

> Change `172.17.8.101` in the example below to an IP address that resolves to a valid web server on your network.

```
$ docker run -d -t \
  --name varnish.1 \
  -p 80:80 \
  --sysctl "net.core.somaxconn=1024" \
  --add-host httpd_1:172.17.8.101 \
  jdeathe/centos-ssh-varnish:2.5.0
```

Verify the named container's process status and health.

```
$ docker ps -a \
  -f "name=varnish.1"
```

Verify successful initialisation of the named container.

```
$ docker logs varnish.1
```

## Instructions

### Running

To run the a docker container from this image you can use the standard docker commands as shown in the example below. Alternatively, there's a [docker-compose](https://github.com/jdeathe/centos-ssh-varnish/blob/centos-7/docker-compose.yml) example.

For production use, it is recommended to select a specific release tag as shown in the examples.

In the following example the http service is bound to port 8000 and offloaded https on port 8500 of the docker host. Also, the environment variable `VARNISH_STORAGE` has been used to set up a 256M memory based storage instead of the default file based type.

#### Using environment variables

```
$ docker stop varnish.1 && \
  docker rm varnish.1; \
  docker run \
  --detach \
  --tty \
  --name varnish.1 \
  --publish 8000:80 \
  --publish 8500:8443 \
  --sysctl "net.core.somaxconn=1024" \
  --sysctl "net.ipv4.ip_local_port_range=1024 65535" \
  --sysctl "net.ipv4.route.flush=1" \
  --ulimit memlock=82000 \
  --ulimit nofile=131072 \
  --ulimit nproc=65535 \
  --env "VARNISH_STORAGE=malloc,256M" \
  --env "VARNISH_MAX_THREADS=2000" \
  --env "VARNISH_MIN_THREADS=100" \
  --add-host httpd_1:172.17.8.101 \
  jdeathe/centos-ssh-varnish:2.5.0
```

#### Environment variables

There are several environmental variables defined at runtime which allows the operator to customise the running container. This may become necessary under special circumstances and the following show those that are most likely to be considered for review, the rest should be left unaltered and for clarification refer to the [varnishd documentation](https://www.varnish-cache.org/docs/6.3/index.html).

##### ENABLE_VARNISHD_WRAPPER

It may be desirable to prevent the startup of the varnishd-wrapper script. For example, when using an image built from this Dockerfile as the source for another Dockerfile you could disable varnishd from startup by setting `ENABLE_VARNISHD_WRAPPER` to `false`.

##### ENABLE_VARNISHNCSA_WRAPPER

Controls the startup of the varnishncsa-wrapper script which is not started by default. With `ENABLE_VARNISHNCSA_WRAPPER` set to `true` the `varnishncsa` process is started to output the Varnish in-memory logs to the log file `/var/log/varnish/access_log`. Logs are in Apache / NCSA combined log format unless altered using `VARNISH_VARNISHNCSA_FORMAT`.

##### VARNISH_MIN_THREADS, VARNISH_MAX_THREADS & VARNISH_THREAD_TIMEOUT

Start at least `VARNISH_MIN_THREADS` but no more than `VARNISH_MAX_THREADS` worker threads with the `VARNISH_THREAD_TIMEOUT` idle timeout.

##### VARNISH_OPTIONS

Use `VARNISH_OPTIONS` to set other `varnishd` options.

##### VARNISH_STORAGE

Use `VARNISH_STORAGE` to specify the storage backend. See the [varnishd documentation](https://varnish-cache.org/docs/6.3/reference/varnishd.html#storage-backend) for the types and parameters available. The default is a file type backend but it is recommended to use malloc if there is enough RAM available.

##### VARNISH_TTL

The `VARNISH_TTL` can be used to set a hard minimum time to live for cached documents. The default is 120 seconds.

##### VARNISH_VARNISHNCSA_FORMAT

When `ENABLE_VARNISHNCSA_WRAPPER` is set to `true` then `VARNISH_VARNISHNCSA_FORMAT` can be used to set the output log [format string](https://varnish-cache.org/docs/6.3/reference/varnishncsa.html#format).

##### VARNISH_VARNISHNCSA_OPTIONS

Use `VARNISH_VARNISHNCSA_OPTIONS` to set other `varnishncsa` options.

##### VARNISH_VCL_CONF

The Varnish VCL configuration file path, (or base64 encoded string of the configuration file contents), is set using `VARNISH_VCL_CONF`. The default configuration supplied is located at the path `/etc/varnish/docker-default.vcl`.


CHANGELOG.md
# Change Log

## 2 - centos-7

Summary of release changes.

### 2.5.0 - 2019-10-10

- Updates Varnish to [6.3.0](https://github.com/varnishcache/varnish-cache/blob/varnish-6.3.0/doc/changes.rst).
- Updates image tags in docker-compose example configuration.

### 2.4.1 - 2019-10-08

- Deprecate Makefile target `logs-delayed`; replaced with `logsdef`.
- Updates Varnish to [6.2.1](https://github.com/varnishcache/varnish-cache/blob/varnish-6.2.1/doc/changes.rst).
- Updates `gcc` package to 4.8.5-39.
- Updates startsecs for `varnishncsa-wrapper` to 5 seconds.
- Updates `test/health_status` helper script with for consistency.
- Updates Makefile target `logs` to accept `[OPTIONS]` (e.g `make -- logs -ft`).
- Updates info/error output for consistency.
- Updates healthcheck failure messages to remove EOL character that is rendered in status response.
- Updates ordering of Tags and respective Dockerfile links in README.md for readability.
- Adds improved test workflow; added `test-setup` target to Makefile.
- Adds Makefile target `logsdef` to handle deferred logs output within a target chain.
- Adds `/docs` directory for supplementary documentation and simplify README.
- Fixes validation failure of 0 second --timeout value in `test/health_status`.

### 2.4.0 - 2019-08-17

- Updates source image to [2.6.0](https://github.com/jdeathe/centos-ssh/releases/tag/2.6.0).
- Updates CHANGELOG.md to simplify maintenance.
- Updates README.md to simplify contents and improve readability.
- Updates README-short.txt to apply to all image variants.
- Updates Dockerfile `org.deathe.description` metadata LABEL for consistency.
- Updates supervisord configuration to send error log output to stderr.
- Updates varnishd supervisord configuration file/priority to `80-varnishd-wrapper.conf`/`80`.
- Updates varnishncsa supervisord configuration file/priority to `50-varnishncsa-wrapper.conf`/`50`.
- Updates docker-compose example with redis session store replacing memcached for the apache-php service.
- Updates wrapper scripts timer to use UTC date timestamps.
- Updates backend probe window from 5 to 3 to reduce time to register an offline backend.
- Fixes docker host connection status check in Makefile.
- Fixes error when restarting/reloading varnishd.
- Adds `inspect`, `reload` and `top` Makefile targets.
- Adds improved `clean` Makefile target; includes exited containers and dangling images.
- Adds `SYSTEM_TIMEZONE` handling to Makefile, scmi, systemd unit and docker-compose templates.
- Adds system time zone validation to healthcheck.
- Adds lock/state file to wrapper scripts.
- Adds VCL to handle `/status` and `/varnish-status` for monitoring the backend and varnish respectively.
- Removes `VARNISH_AUTOSTART_VARNISHD_WRAPPER`, replaced with `ENABLE_VARNISHD_WRAPPER`.
- Removes `VARNISH_AUTOSTART_VARNISHNCSA_WRAPPER`, replaced with `ENABLE_VARNISHNCSA_WRAPPER`.
- Removes support for long image tags (i.e. centos-7-2.x.x).

### 2.3.0 - 2019-05-06

- Updates `gcc` package to gcc-4.8.5-36.el7_6.1.
- Updates source image to [2.5.1](https://github.com/jdeathe/centos-ssh/releases/tag/2.5.1).
- Updates Varnish to [6.2.0](https://github.com/varnishcache/varnish-cache/blob/varnish-6.2.0/doc/changes.rst).
- Updates and restructures Dockerfile.
- Updates container naming conventions and readability of `Makefile`.
- Updates startup time to 4 seconds.
- Updates healthcheck retries to 5.
- Updates default VCL excluding several parts already defined in `builtin.vcl`.
- Fixes issue with unexpected published port in run templates when `DOCKER_PORT_MAP_TCP_80` or `DOCKER_PORT_MAP_TCP_8443` is set to an empty string or 0.
- Fixes binary paths in systemd unit files for compatibility with both EL and Ubuntu hosts.
- Adds consideration for event lag into test cases for unhealthy health_status events.
- Adds port incrementation to Makefile's run template for container names with an instance suffix.
- Adds placeholder replacement of `RELEASE_VERSION` docker argument to systemd service unit template.
- Adds improvement to pull logic in systemd unit install template.
- Adds `SSH_AUTOSTART_SUPERVISOR_STDOUT` with a value "false", disabling startup of `supervisor_stdout`.
- Adds error messages to healthcheck script and includes supervisord check.
- Adds improved logging output.
- Adds docker-compose configuration example.
- Adds improved/simplified Cookie logic in `docker-default.vcl`.
- Adds improved `healtchcheck`, `varnishd-wrapper` and `varnishncsa-wrapper` scripts.
- Adds improved lock/state file implementation in wrapper scripts.
- Adds `VARNISH_OPTIONS` and `VARNISH_VARNISHNCSA_OPTIONS`.
- Adds improved VCL error checking/handling.
- Adds styled synthetic 500 error responses.
- Removes use of `/etc/services-config` paths.
- Removes the unused group element from the default container name.
- Removes the node element from the default container name.
- Removes unused environment variables from Makefile and scmi configuration.
- Removes X-Fleet section from etcd register template unit-file.

### 2.2.1 - 2018-12-10

- Fixes typo in test; using `--format` instead of `--filter`.
- Updates source image to [2.4.1](https://github.com/jdeathe/centos-ssh/releases/tag/2.4.1).
- Updates Varnish to [6.1.1](https://github.com/varnishcache/varnish-cache/blob/varnish-6.1.1/doc/changes.rst).
- Updates `gcc` packages to 4.8.5-36.
- Updates image versions in docker-compose example and tests.
- Adds required `--sysctl` settings to docker run templates.
- Adds change to ensure varnishncsa is run with a non-root user `varnishlog`.
- Adds varnishncsa access logs to docker log output.
- Adds "Varnish Details" to docker log output.

### 2.2.0 - 2018-10-09

- Updates Varnish to [6.1.0](https://github.com/varnishcache/varnish-cache/blob/varnish-6.1.0/doc/changes.rst).
- Adds lockfile to ensure varnishd is started before running varnishncsa.
- Adds failure messages to healthcheck script.

### 2.1.0 - 2018-10-01


test/fixture/apache/var/www/public_html/session.php
<?php
	session_start();
	$_SESSION['integer'] = 123;
	$_SESSION['float'] = 12345.67890;
	$_SESSION['string'] = '@string:#\$£';
	session_write_close();
	var_dump($_SESSION);


run.sh
#!/usr/bin/env bash

DIR_PATH="$( if [ "$( echo "${0%/*}" )" != "$( echo "${0}" )" ] ; then cd "$( echo "${0%/*}" )"; fi; pwd )"
if [[ $DIR_PATH == */* ]] && [[ $DIR_PATH != "$( pwd )" ]] ; then
	cd $DIR_PATH
fi

source run.conf

have_docker_container_name ()
{
	local NAME=$1

	if [[ -n $(docker ps -a | awk -v pattern="^${NAME}$" '$NF ~ pattern { print $NF; }') ]]; then
		return 0
	else
		return 1
	fi
}

is_docker_container_name_running ()
{
	local NAME=$1

	if [[ -n $(docker ps | awk -v pattern="^${NAME}$" '$NF ~ pattern { print $NF; }') ]]; then
		return 0
	else
		return 1
	fi
}

remove_docker_container_name ()
{
	local NAME=$1

	if have_docker_container_name ${NAME} ; then
		if is_docker_container_name_running ${NAME} ; then
			echo Stopping container ${NAME}...
			(docker stop ${NAME})
		fi
		echo Removing container ${NAME}...
		(docker rm ${NAME})
	fi
}

# Set the --add-host parameters
DOCKER_HOST_TYPE=${DOCKER_HOST_TYPE:-default}
ADD_BACKEND_HOSTS=

case ${DOCKER_HOST_TYPE} in
	cluster)
		if [[ ${VARNISH_VCL_CONF} == /etc/varnish/docker-cluster.vcl ]]; then
			echo Running Varnish with 3 backend cluster nodes.
			ADD_BACKEND_HOSTS="--add-host backend-1:${DOCKER_HOST_IP_CLUSTER_01} \
								--add-host backend-2:${DOCKER_HOST_IP_CLUSTER_02} \
								--add-host backend-3:${DOCKER_HOST_IP_CLUSTER_03}"
		else
			echo Running Varnish with 1 backend cluster node.
			ADD_BACKEND_HOSTS="--add-host backend-1:${DOCKER_HOST_IP_CLUSTER_01}"
		fi
		;;
	*)
		echo Running Varnish with 1 backend node.

		# Fail if attempting to use the wrong Varnish VCL configuration.
		if [[ ${VARNISH_VCL_CONF} == /etc/varnish/docker-cluster.vcl ]]; then
			echo ERROR: Varnish configuration docker-cluster.vcl requires DOCKER_HOST_TYPE=cluster
			exit 1
		fi

		DOCKER_HOST_IP=
		if [[ ${DOCKER_HOST} != "" ]]; then
			DOCKER_HOST_IP=$(echo ${DOCKER_HOST} | grep -oE "[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}")
		fi

		if [[ ${DOCKER_HOST_IP} != "" ]] && [[ ${DOCKER_HOST_IP} != ${DOCKER_HOST_IP_DEFAULT} ]]; then
			echo Found non-standard DOCKER_HOST defined.
			echo Using ${DOCKER_HOST_IP} instead of ${DOCKER_HOST_IP_DEFAULT} for backend-1.
			ADD_BACKEND_HOSTS="--add-host backend-1:${DOCKER_HOST_IP}"
		else
			ADD_BACKEND_HOSTS="--add-host backend-1:${DOCKER_HOST_IP_DEFAULT}"
		fi
		;;
esac

# Configuration volume
if ! have_docker_container_name ${VOLUME_CONFIG_NAME} ; then

	# For configuration that is specific to the running container
	CONTAINER_MOUNT_PATH_CONFIG=${MOUNT_PATH_CONFIG}/${DOCKER_NAME}

	# For configuration that is shared across a group of containers
	CONTAINER_MOUNT_PATH_CONFIG_SHARED_SSH=${MOUNT_PATH_CONFIG}/ssh.${SERVICE_UNIT_SHARED_GROUP}

	if [ ! -d ${CONTAINER_MOUNT_PATH_CONFIG_SHARED_SSH}/ssh ]; then
		CMD=$(mkdir -p ${CONTAINER_MOUNT_PATH_CONFIG_SHARED_SSH}/ssh)
		$CMD || sudo $CMD
	fi

	# Configuration for SSH is from jdeathe/centos-ssh/etc/services-config/ssh
	#if [[ ! -n $(find ${CONTAINER_MOUNT_PATH_CONFIG_SHARED_SSH}/ssh -maxdepth 1 -type f) ]]; then
	#		CMD=$(cp -R etc/services-config/ssh/ ${CONTAINER_MOUNT_PATH_CONFIG_SHARED_SSH}/ssh/)
	#		$CMD || sudo $CMD
	#fi

	if [ ! -d ${CONTAINER_MOUNT_PATH_CONFIG}/supervisor ]; then
		CMD=$(mkdir -p ${CONTAINER_MOUNT_PATH_CONFIG}/supervisor)
		$CMD || sudo $CMD
	fi

	if [[ ! -n $(find ${CONTAINER_MOUNT_PATH_CONFIG}/supervisor -maxdepth 1 -type f) ]]; then
		CMD=$(cp -R etc/services-config/supervisor ${CONTAINER_MOUNT_PATH_CONFIG}/)
		$CMD || sudo $CMD
	fi

	if [ ! -d ${CONTAINER_MOUNT_PATH_CONFIG}/varnish ]; then
		CMD=$(mkdir -p ${CONTAINER_MOUNT_PATH_CONFIG}/varnish)
		$CMD || sudo $CMD
	fi

	if [[ ! -n $(find ${CONTAINER_MOUNT_PATH_CONFIG}/varnish -maxdepth 1 -type f) ]]; then
		CMD=$(cp -R etc/services-config/varnish ${CONTAINER_MOUNT_PATH_CONFIG}/)
		$CMD || sudo $CMD
	fi
(
set -x
docker run \
	--name ${VOLUME_CONFIG_NAME} \
	-v ${CONTAINER_MOUNT_PATH_CONFIG_SHARED_SSH}/ssh:/etc/services-config/ssh \
	-v ${CONTAINER_MOUNT_PATH_CONFIG}/supervisor:/etc/services-config/supervisor \
	-v ${CONTAINER_MOUNT_PATH_CONFIG}/varnish:/etc/services-config/varnish \
	busybox:latest \
	/bin/true;
)
fi

# Force replace container of same name if found to exist
remove_docker_container_name ${DOCKER_NAME}

# In a sub-shell set xtrace - prints the docker command to screen for reference
(
set -x
docker run \
	-d \
	--privileged \
	--name ${DOCKER_NAME} \
	-p 8000:80 \
	-p 8500:8443 \
	--env "VARNISH_VCL_CONF=${VARNISH_VCL_CONF}" \
	--env "VARNISH_STORAGE=${VARNISH_STORAGE}" \
	${ADD_BACKEND_HOSTS} \
	--volumes-from ${VOLUME_CONFIG_NAME} \
	${DOCKER_IMAGE_REPOSITORY_NAME}
)

if is_docker_container_name_running ${DOCKER_NAME} ; then
	docker ps | grep -v -e "${DOCKER_NAME}/.*,.*" | grep ${DOCKER_NAME}
	echo " ---> Docker container running."
fi


systemd-install.sh
#!/usr/bin/env bash

DIR_PATH="$( if [ "$( echo "${0%/*}" )" != "$( echo "${0}" )" ] ; then cd "$( echo "${0%/*}" )"; fi; pwd )"
if [[ $DIR_PATH == */* ]] && [[ $DIR_PATH != "$( pwd )" ]] ; then
	cd $DIR_PATH
fi

COREOS_PRIVATE_IPV4=${COREOS_PRIVATE_IPV4:-$(grep 'COREOS_PRIVATE_IPV4=' /etc/environment | sed 's~COREOS_PRIVATE_IPV4=~~g')}

have_docker_container_name ()
{
	local NAME=$1

	if [[ -n $(docker ps -a | awk -v pattern="^${NAME}$" '$NF ~ pattern { print $NF; }') ]]; then
		return 0
	else
		return 1
	fi
}

is_docker_container_name_running ()
{
	local NAME=$1

	if [[ -n $(docker ps | awk -v pattern="^${NAME}$" '$NF ~ pattern { print $NF; }') ]]; then
		return 0
	else
		return 1
	fi
}

remove_docker_container_name ()
{
	local NAME=$1

	if have_docker_container_name ${NAME} ; then
		if is_docker_container_name_running ${NAME} ; then
			echo Stopping container ${NAME}
			(docker stop ${NAME})
		fi
		echo Removing container ${NAME}
		(docker rm ${NAME})
	fi
}

loading_counter ()
{
	local COUNTER=${1:-10}

	while [ ${COUNTER} -ge 1 ]; do
		echo -ne "Loading in: ${COUNTER} \r"
		sleep 1
		COUNTER=$[${COUNTER}-1]
	done
}

OPT_SERVICE_NAME_FULL=${SERVICE_NAME_FULL:-varnish.pool-1.1.1@8000.service}
OPT_SERVICE_NAME_SHORT=$(cut -d '@' -f1 <<< "${OPT_SERVICE_NAME_FULL}")

# Add required configuration directories
mkdir -p /etc/services-config/${OPT_SERVICE_NAME_SHORT}/{varnish,supervisor}

if [[ ! -n $(find /etc/services-config/${OPT_SERVICE_NAME_SHORT}/supervisor -maxdepth 1 -type f) ]]; then
	cp -R etc/services-config/supervisor /etc/services-config/${OPT_SERVICE_NAME_SHORT}/
fi

if [[ ! -n $(find /etc/services-config/${OPT_SERVICE_NAME_SHORT}/varnish -maxdepth 1 -type f) ]]; then
	cp -R etc/services-config/varnish /etc/services-config/${OPT_SERVICE_NAME_SHORT}/
fi

# Stop remove any existing application and configuration volume containers.
sudo systemctl stop ${OPT_SERVICE_NAME_FULL} &> /dev/null
remove_docker_container_name volume-config.${OPT_SERVICE_NAME_SHORT}
remove_docker_container_name ${OPT_SERVICE_NAME_SHORT}

sudo cp ${OPT_SERVICE_NAME_FULL} /etc/systemd/system/
if [[ ${COREOS_PRIVATE_IPV4} != "" ]]; then
	sudo sed -i "s~ backend-1:172.17.8.101~ backend-1:${COREOS_PRIVATE_IPV4}~g" /etc/systemd/system/${OPT_SERVICE_NAME_FULL}
fi
sudo systemctl daemon-reload
sudo systemctl enable /etc/systemd/system/${OPT_SERVICE_NAME_FULL}

echo "WARNING: This may take a while if pulling large container images for the first time..."
sudo systemctl restart ${OPT_SERVICE_NAME_FULL}

loading_counter 3

docker logs ${OPT_SERVICE_NAME_SHORT}

src/opt/scmi/default.sh

# Handle incrementing the docker host port for instances unless a port range is defined.
DOCKER_PUBLISH=
if [[ ${DOCKER_PORT_MAP_TCP_80} != NULL ]]
then
	if grep -qE \
			'^([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}:)?[1-9][0-9]*$' \
			<<< "${DOCKER_PORT_MAP_TCP_80}" \
		&& grep -qE \
			'^.+\.[0-9]+(\.[0-9]+)?$' \
			<<< "${DOCKER_NAME}"
	then
		printf -v \
			DOCKER_PUBLISH \
			-- '%s --publish %s%s:80' \
			"${DOCKER_PUBLISH}" \
			"$(
				grep -o \
					'^[0-9\.]*:' \
					<<< "${DOCKER_PORT_MAP_TCP_80}"
			)" \
			"$(( 
				$(
					grep -oE \
						'[0-9]+$' \
						<<< "${DOCKER_PORT_MAP_TCP_80}"
				) \
				+ $(
					grep -oE \
						'([0-9]+)(\.[0-9]+)?$' \
						<<< "${DOCKER_NAME}" \
					| awk -F. \
						'{ print $1; }'
				) \
				- 1
			))"
	else
		printf -v \
			DOCKER_PUBLISH \
			-- '%s --publish %s:80' \
			"${DOCKER_PUBLISH}" \
			"${DOCKER_PORT_MAP_TCP_80}"
	fi
fi

if [[ ${DOCKER_PORT_MAP_TCP_8443} != NULL ]]
then
	if grep -qE \
			'^([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}:)?[1-9][0-9]*$' \
			<<< "${DOCKER_PORT_MAP_TCP_8443}" \
		&& grep -qE \
			'^.+\.[0-9]+(\.[0-9]+)?$' \
			<<< "${DOCKER_NAME}"
	then
		printf -v \
			DOCKER_PUBLISH \
			-- '%s --publish %s%s:8443' \
			"${DOCKER_PUBLISH}" \
			"$(
				grep -o \
					'^[0-9\.]*:' \
					<<< "${DOCKER_PORT_MAP_TCP_8443}"
			)" \
			"$(( 
				$(
					grep -oE \
						'[0-9]+$' \
						<<< "${DOCKER_PORT_MAP_TCP_8443}"
				) \
				+ $(
					grep -oE \
						'([0-9]+)(\.[0-9]+)?$' \
						<<< "${DOCKER_NAME}" \
					| awk -F. \
						'{ print $1; }'
				) \
				- 1
			))"
	else
		printf -v \
			DOCKER_PUBLISH \
			-- '%s --publish %s:8443' \
			"${DOCKER_PUBLISH}" \
			"${DOCKER_PORT_MAP_TCP_8443}"
	fi
fi

# Common parameters of create and run targets
DOCKER_CONTAINER_PARAMETERS="--tty \
--name ${DOCKER_NAME} \
--restart ${DOCKER_RESTART_POLICY} \
--sysctl \"net.core.somaxconn=${SYSCTL_NET_CORE_SOMAXCONN}\" \
--sysctl \"net.ipv4.ip_local_port_range=${SYSCTL_NET_IPV4_IP_LOCAL_PORT_RANGE}\" \
--sysctl \"net.ipv4.route.flush=${SYSCTL_NET_IPV4_ROUTE_FLUSH}\" \
--ulimit \"memlock=${ULIMIT_MEMLOCK}\" \
--ulimit \"nofile=${ULIMIT_NOFILE}\" \
--ulimit \"nproc=${ULIMIT_NPROC}\" \
--env \"ENABLE_VARNISHD_WRAPPER=${ENABLE_VARNISHD_WRAPPER}\" \
--env \"ENABLE_VARNISHNCSA_WRAPPER=${ENABLE_VARNISHNCSA_WRAPPER}\" \
--env \"SYSTEM_TIMEZONE=${SYSTEM_TIMEZONE}\" \
--env \"VARNISH_MAX_THREADS=${VARNISH_MAX_THREADS}\" \
--env \"VARNISH_MIN_THREADS=${VARNISH_MIN_THREADS}\" \
--env \"VARNISH_OPTIONS=${VARNISH_OPTIONS}\" \
--env \"VARNISH_STORAGE=${VARNISH_STORAGE}\" \
--env \"VARNISH_THREAD_TIMEOUT=${VARNISH_THREAD_TIMEOUT}\" \
--env \"VARNISH_TTL=${VARNISH_TTL}\" \
--env \"VARNISH_VARNISHNCSA_FORMAT=${VARNISH_VARNISHNCSA_FORMAT}\" \
--env \"VARNISH_VARNISHNCSA_OPTIONS=${VARNISH_VARNISHNCSA_OPTIONS}\" \
--env \"VARNISH_VCL_CONF=${VARNISH_VCL_CONF}\" \
${DOCKER_PUBLISH}"


src/opt/scmi/service-unit.sh
# ------------------------------------------------------------------------------
# Constants
# ------------------------------------------------------------------------------
readonly SERVICE_UNIT_ENVIRONMENT_KEYS="
 DOCKER_CONTAINER_OPTS
 DOCKER_IMAGE_PACKAGE_PATH
 DOCKER_IMAGE_TAG
 DOCKER_PORT_MAP_TCP_80
 DOCKER_PORT_MAP_TCP_8443
 ENABLE_VARNISHD_WRAPPER
 ENABLE_VARNISHNCSA_WRAPPER
 SYSCTL_NET_CORE_SOMAXCONN
 SYSCTL_NET_IPV4_IP_LOCAL_PORT_RANGE
 SYSCTL_NET_IPV4_ROUTE_FLUSH
 SYSTEM_TIMEZONE
 ULIMIT_MEMLOCK
 ULIMIT_NOFILE
 ULIMIT_NPROC
 VARNISH_MAX_THREADS
 VARNISH_MIN_THREADS
 VARNISH_OPTIONS
 VARNISH_STORAGE
 VARNISH_THREAD_TIMEOUT
 VARNISH_TTL
 VARNISH_VARNISHNCSA_FORMAT
 VARNISH_VARNISHNCSA_OPTIONS
 VARNISH_VCL_CONF
"

readonly SERVICE_UNIT_REGISTER_ENVIRONMENT_KEYS="
 REGISTER_ETCD_PARAMETERS
 REGISTER_TTL
 REGISTER_UPDATE_INTERVAL
"

# ------------------------------------------------------------------------------
# Variables
# ------------------------------------------------------------------------------
SERVICE_UNIT_INSTALL_TIMEOUT="${SERVICE_UNIT_INSTALL_TIMEOUT:-7}"


src/opt/scmi/environment.sh
# ------------------------------------------------------------------------------
# Constants
# ------------------------------------------------------------------------------
readonly DOCKER_IMAGE_NAME=centos-ssh-varnish
readonly DOCKER_IMAGE_RELEASE_TAG_PATTERN='^[1-2]\.[0-9]+\.[0-9]+$'
readonly DOCKER_IMAGE_TAG_PATTERN='^(latest|[1-2]\.[0-9]+\.[0-9]+)$'
readonly DOCKER_USER=jdeathe

# ------------------------------------------------------------------------------
# Variables
# ------------------------------------------------------------------------------
DIST_PATH="${DIST_PATH:-./dist}"
DOCKER_CONTAINER_OPTS="${DOCKER_CONTAINER_OPTS:-}"
DOCKER_IMAGE_TAG="${DOCKER_IMAGE_TAG:-latest}"
DOCKER_NAME="${DOCKER_NAME:-varnish.1}"
DOCKER_PORT_MAP_TCP_80="${DOCKER_PORT_MAP_TCP_80:-8000}"
DOCKER_PORT_MAP_TCP_8443="${DOCKER_PORT_MAP_TCP_8443:-8500}"
DOCKER_RESTART_POLICY="${DOCKER_RESTART_POLICY:-always}"
NO_CACHE="${NO_CACHE:-false}"
REGISTER_ETCD_PARAMETERS="${REGISTER_ETCD_PARAMETERS:-}"
REGISTER_TTL="${REGISTER_TTL:-60}"
REGISTER_UPDATE_INTERVAL="${REGISTER_UPDATE_INTERVAL:-55}"
STARTUP_TIME="${STARTUP_TIME:-4}"
SYSCTL_NET_CORE_SOMAXCONN="${SYSCTL_NET_CORE_SOMAXCONN:-1024}"
SYSCTL_NET_IPV4_IP_LOCAL_PORT_RANGE="${SYSCTL_NET_IPV4_IP_LOCAL_PORT_RANGE:-"1024 65535"}"
SYSCTL_NET_IPV4_ROUTE_FLUSH="${SYSCTL_NET_IPV4_ROUTE_FLUSH:-1}"
ULIMIT_MEMLOCK="${ULIMIT_MEMLOCK:-82000}"
ULIMIT_NOFILE="${ULIMIT_NOFILE:-131072}"
ULIMIT_NPROC="${ULIMIT_NPROC:-9223372036854775807}"

# ------------------------------------------------------------------------------
# Application container configuration
# ------------------------------------------------------------------------------
ENABLE_VARNISHD_WRAPPER="${ENABLE_VARNISHD_WRAPPER:-true}"
ENABLE_VARNISHNCSA_WRAPPER="${ENABLE_VARNISHNCSA_WRAPPER:-false}"
SYSTEM_TIMEZONE="${SYSTEM_TIMEZONE:-UTC}"
VARNISH_MAX_THREADS="${VARNISH_MAX_THREADS:-1000}"
VARNISH_MIN_THREADS="${VARNISH_MIN_THREADS:-50}"
VARNISH_OPTIONS=""
VARNISH_STORAGE="${VARNISH_STORAGE:-file,/var/lib/varnish/varnish_storage.bin,1G}"
VARNISH_THREAD_TIMEOUT="${VARNISH_THREAD_TIMEOUT:-120}"
VARNISH_TTL="${VARNISH_TTL:-120}"
VARNISH_VARNISHNCSA_FORMAT="${VARNISH_VARNISHNCSA_FORMAT:-"%h %l %u %t \"%r\" %s %b \"%{Referer}i\" \"%{User-agent}i\""}"
VARNISH_VARNISHNCSA_OPTIONS=""
VARNISH_VCL_CONF="${VARNISH_VCL_CONF:-/etc/varnish/docker-default.vcl}"
